### ISSUE TITLE
IL2CPP incompatibility.

### ISSUE THREAD

**Issue by D3m0n92** [pusher/pusher-websocket-unity#15]
As soon as the app starts, I get the following error:

NotSupportedException: System.Reflection.Emit.DynamicMethod::.ctor

System.Reflection.Emit.DynamicMethod..ctor (System.String name, System.Type returnType, System.Type[] parameterTypes, System.Type owner, System.Boolean skipVisibility) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.Utilities.DynamicReflectionDelegateFactory.CreateDynamicMethod (System.String name, System.Type returnType, System.Type[] parameterTypes, System.Type owner) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.Utilities.DynamicReflectionDelegateFactory.CreateParameterizedConstructor (System.Reflection.MethodBase method) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.Serialization.DefaultContractResolver.CreateObjectContract (System.Type objectType) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.Serialization.DefaultContractResolver.CreateContract (System.Type objectType) (at <00000000000000000000000000000000>:0)
System.Func`2[T,TResult].Invoke (T arg) (at <00000000000000000000000000000000>:0)
System.Collections.Concurrent.ConcurrentDictionary`2[TKey,TValue].GetOrAdd (TKey key, System.Func`2[T,TResult] valueFactory) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.Utilities.ThreadSafeStore`2[TKey,TValue].Get (TKey key) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.Serialization.DefaultContractResolver.ResolveContract (System.Type type) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.Serialization.JsonSerializerInternalReader.GetContractSafe (System.Type type) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize (Newtonsoft.Json.JsonReader reader, System.Type objectType, System.Boolean checkAdditionalContent) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.JsonSerializer.DeserializeInternal (Newtonsoft.Json.JsonReader reader, System.Type objectType) (at <00000000000000000000000000000000>:0)
Newtonsoft.Json.JsonSerializer.Deserialize (Newtonsoft.Json.Json

**Comment by damdo** [comment]
Ciao Vincenzo,
I'm sorry you are running into this issue and thank you for reporting it.
This is an issue caused by the fact that [pusher-websocket-dotnet](https://github.com/pusher/pusher-websocket-dotnet) (the parent library that pusher-websocket-unity is using) utilises [Newtonsoft.Json](https://github.com/JamesNK/Newtonsoft.Json) which is still incompatible with IL2CPP (here is the PR that could potentially make it compatible https://github.com/JamesNK/Newtonsoft.Json/pull/2044). More about that here: (https://github.com/pusher/pusher-websocket-dotnet/issues/83).

We are considering different options to fix this, and it is for us high priority to do so.
Please bear with us, we will keep you posted.

**Comment by twomedia** [comment]
Glad to see an official Unity lib for Pusher! We're waiting on IL2CPP support as well before we can move away from our custom implementation. Hopefully we'll see that PR merged soon 🤞🏻

**Comment by damdo** [comment]
Hey Thomas (@twomedia), thank you.

I've just released a new version of **pusher-websocket-unity** 
here -> https://github.com/pusher/pusher-websocket-unity/releases/tag/1.1.2%2B200211 

This version switches to the fork of Newtonsoft.Json with support for IL2CPP that I mentioned in the comment above (https://github.com/jilleJr/Newtonsoft.Json-for-Unity).

It would be great if you could give some feedback on it, and see if it works for you 😬

**Comment by twomedia** [comment]
@damdo Thanks for jumping on this so quickly! Unfortunately, we haven't had much luck getting it working in both our projects today. We seem to be getting the following error spammed the the console when attempting to connect:
```
Pusher Error: Missing parameter: event
PusherManager:OnPusherOnError(Object, PusherException)
PusherClient.Connection:ParseError(String)
WebSocket4Net.WebSocket:ExecuteCommand(WebSocketCommandInfo)
WebSocket4Net.WebSocket:OnDataReceived(Byte[], Int32, Int32)
SuperSocket.ClientEngine.AuthenticatedStreamTcpSession:OnDataRead(IAsyncResult)
System.Threading.Tasks.AwaitTaskContinuation:InvokeAction(Object)
System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
System.Threading.Tasks.AwaitTaskContinuation:RunCallback(ContextCallback, Object, Task&)
System.Threading.Tasks.Task:FinishContinuations()
System.Threading.Tasks.Task`1:TrySetResult(TResult)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:Create()
Mono.Net.Security.<>c__DisplayClass66_0:<InnerRead>b__0()
System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
System.Runtime.CompilerServices.MoveNextRunner:Run()
System.Threading.Tasks.AwaitTaskContinuation:RunOrScheduleAction(Action, Boolean, Task&)
System.Threading.Tasks.Task:FinishContinuations()
System.Threading.Tasks.Task`1:TrySetResult(TResult)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:Create()
Mono.Net.Security.MobileAuthenticatedStream:InnerWrite(Boolean, CancellationToken)
System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
System.Runtime.CompilerServices.MoveNextRunner:Run()
System.Threading.Tasks.AwaitTaskContinuation:RunOrScheduleAction(Action, Boolean, Task&)
System.Threading.Tasks.Task:FinishContinuations()
System.Threading.Tasks.Task`1:TrySetResult(TResult)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:Create()
Mono.Net.Security.MobileAuthenticatedStream:InnerRead(Boolean, Int32, CancellationToken)
System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
System.Runtime.CompilerServices.MoveNextRunner:Run()
System.Threading.Tasks.AwaitTaskContinuation:RunOrScheduleAction(Action, Boolean, Task&)
System.Threading.Tasks.Task:FinishContinuations()
System.Threading.Tasks.Task`1:TrySetResult(TResult)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:Create()
Mono.Net.Security.AsyncProtocolRequest:ToString()
System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
System.Runtime.CompilerServices.MoveNextRunner:Run()
System.Threading.Tasks.AwaitTaskContinuation:RunOrScheduleAction(Action, Boolean, Task&)
System.Threading.Tasks.Task:FinishContinuations()
System.Threading.Tasks.Task`1:TrySetResult(TResult)
System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1:Create()
Mono.Net.Security.<>c__DisplayClass66_0:<InnerRead>b__0()
System.Runtime.CompilerServices.MoveNextRunner:InvokeMoveNext(Object)
System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
System.Runtime.CompilerServices.MoveNextRunner:Run()
System.Threading.Tasks.AwaitTaskContinuation:RunOrScheduleAction(Action, Boolean, Task&)
System.Threading.Tasks.Task:FinishContinuations()
System.Threading.Tasks.Task`1:TrySetResult(TResult)
System.Threading.Tasks.FromAsyncTrimPromise`1:Complete(TInstance, Func`3, IAsyncResult, Boolean)
System.Net.Sockets.<>c:<Complete>b__27_0(Object)
System.Threading.ThreadPoolWorkQueue:Dispatch()
```
Ultimately this ends up causing Pusher to switch between `Initialized -> Disconnecting -> Disconnected -> WaitingToReconnect` states. 

In another unrelated issue, when we tried installing the lib using Unity Package Manager we got the following error:
<img width="622" alt="Screen Shot 2020-02-12 at 10 19 34 am" src="https://user-images.githubusercontent.com/1635048/74321841-155dd600-4dd7-11ea-9dc5-8f7098c6f736.png">
We ended up importing using the .unitypackage with no issues.

Sorry for the long post and thanks in advance for the help!

**Comment by twomedia** [comment]
Just an update on this issue, we're also seeing the "Pusher Error: Missing parameter: event" error appearing in the Pusher.com error logs page 
<img width="888" alt="Screen Shot 2020-02-26 at 4 13 47 pm" src="https://user-images.githubusercontent.com/1635048/75314231-42b88280-58b3-11ea-831b-b008d752e0ee.png">

**Comment by CliffCawley** [comment]
I was able to fix this error by adding a `link.xml` with the contents:
```
<linker>
       <assembly fullname="PusherClient" preserve="all"/>
</linker>
```

**Comment by twomedia** [comment]
Just an update on the above. Thanks @CliffCawley, adding to the `link.xml` file partially fixed the issue for us. We did however have another issue where the app would crash when receiving a channel event. Further investigation found that this was caused by us using the dynamic channel.Bind() method instead of either the string or PusherEvent overloads (As the dynamic keyword is not support by IL2CPP due to its AOT requirements). We seem to be stable now on both of our products and we're entering public testing soon.
